File: 03.0-very-simple-http-server/server.2.recv-replaced.rb
require 'socket' # A library built into Ruby - provides the TCPServer class

$stdout.sync = 1  # Line-buffer output to STDOUT.
listener = TCPServer.open(8080)
loop do
  socket = listener.accept  # Wait til a client connects, then open a socket.
  puts "FROM THE CLIENT: "
  while txt = socket.readline
    puts txt
    if txt == "\r\n"  # i.e., CRLF
      puts "reached end of request"
      socket.puts("Hello, world")
      puts "SERVER: Sent 'Hello, world' response to client"
      socket.close
      break
    end
  end
end
puts "SERVER: now I'm exiting"

File: 03.0-very-simple-http-server/server.2.recv-replaced.py
import socket
import os
import sys
from unbuffered_reading import read_line_unbuffered_from

sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)  # Line-buffer output to STDOUT.
listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
listener.bind(('0.0.0.0', 8080))
listener.listen(5)
while 1:
  (socket,address) = listener.accept()
  print("FROM THE CLIENT: ")
  while 1:
    txt = read_line_unbuffered_from(socket)
    if len(txt)==0:
      break
    print(txt)
    if txt == "\r\n":  # i.e., CRLF
      print("SERVER: reached end of request headers")
      socket.send("Hello, world\n".encode('utf8'))
      print("SERVER: Sent 'Hello, world' response to client")
      socket.close()
      break
print("SERVER: now I'm exiting")
